Reimplement c10::apply / c10::invoke with c++14

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30924 Reimplement c10::apply / c10::invoke with c++14**

These had an incomplete implementation before because the one from cppreference.com didn't work in C++11.
Now with C++14, we can make the full implementation from cppreference.com work and have c++17-compatible verisons of std::apply/std::invoke.

Differential Revision: [D18869642](https://our.internmc.facebook.com/intern/diff/D18869642/)

smessmer committed:
Reimplement c10::apply / c10::invoke with c++14

These had an incomplete implementation before because the one from cppreference.com didn't work in C++11.
Now with C++14, we can make the full implementation from cppreference.com work and have c++17-compatible verisons of std::apply/std::invoke.

Differential Revision: [D18869642](https://our.internmc.facebook.com/intern/diff/D18869642/)

[ghstack-poisoned]

ezyang reviewed (approved):
Did not review, taking on faith that you faithfully copied cppreference

smessmer committed:
Update on "Reimplement c10::apply / c10::invoke with c++14"


These had an incomplete implementation before because the one from cppreference.com didn't work in C++11.
Now with C++14, we can make the full implementation from cppreference.com work and have c++17-compatible verisons of std::apply/std::invoke.

Differential Revision: [D18869642](https://our.internmc.facebook.com/intern/diff/D18869642/)

[ghstack-poisoned]

smessmer committed:
Update on "Reimplement c10::apply / c10::invoke with c++14"


These had an incomplete implementation before because the one from cppreference.com didn't work in C++11.
Now with C++14, we can make the full implementation from cppreference.com work and have c++17-compatible verisons of std::apply/std::invoke.

Differential Revision: [D18869642](https://our.internmc.facebook.com/intern/diff/D18869642/)

[ghstack-poisoned]

smessmer committed:
Update on "Reimplement c10::apply / c10::invoke with c++14"


These had an incomplete implementation before because the one from cppreference.com didn't work in C++11.
Now with C++14, we can make the full implementation from cppreference.com work and have c++17-compatible verisons of std::apply/std::invoke.

Differential Revision: [D18869642](https://our.internmc.facebook.com/intern/diff/D18869642/)

[ghstack-poisoned]

smessmer committed:
Update on "Reimplement c10::apply / c10::invoke with c++14"


These had an incomplete implementation before because the one from cppreference.com didn't work in C++11.
Now with C++14, we can make the full implementation from cppreference.com work and have c++17-compatible verisons of std::apply/std::invoke.

Differential Revision: [D18869642](https://our.internmc.facebook.com/intern/diff/D18869642/)

[ghstack-poisoned]

smessmer committed:
Update on "Reimplement c10::apply / c10::invoke with c++14"


These had an incomplete implementation before because the one from cppreference.com didn't work in C++11.
Now with C++14, we can make the full implementation from cppreference.com work and have c++17-compatible verisons of std::apply/std::invoke.

Differential Revision: [D18869642](https://our.internmc.facebook.com/intern/diff/D18869642/)

[ghstack-poisoned]

smessmer committed:
Update on "Reimplement c10::apply / c10::invoke with c++14"


These had an incomplete implementation before because the one from cppreference.com didn't work in C++11.
Now with C++14, we can make the full implementation from cppreference.com work and have c++17-compatible verisons of std::apply/std::invoke.

Differential Revision: [D18869642](https://our.internmc.facebook.com/intern/diff/D18869642/)

[ghstack-poisoned]

smessmer committed:
Update on "Reimplement c10::apply / c10::invoke with c++14"


These had an incomplete implementation before because the one from cppreference.com didn't work in C++11.
Now with C++14, we can make the full implementation from cppreference.com work and have c++17-compatible verisons of std::apply/std::invoke.

Differential Revision: [D18869642](https://our.internmc.facebook.com/intern/diff/D18869642/)

[ghstack-poisoned]

smessmer committed:
Update on "Reimplement c10::apply / c10::invoke with c++14"


These had an incomplete implementation before because the one from cppreference.com didn't work in C++11.
Now with C++14, we can make the full implementation from cppreference.com work and have c++17-compatible verisons of std::apply/std::invoke.

Differential Revision: [D18869642](https://our.internmc.facebook.com/intern/diff/D18869642/)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit e0dd334:
* **5/5** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=e0dd33482447baf7b46b6ab86db7d637094c5115).

### 5 new failures recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3962138) caffe2_onnx_py2_gcc5_ubuntu16_04_build (1/5)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74950126))
<details>
<summary>
<code>Dec 16 10:01:38 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected string-literal before ')' token</code>
</summary>

```
Dec 16 10:01:38                                                                  ^ 
Dec 16 10:01:38 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:105:65: error: expected string-literal before ')' token 
Dec 16 10:01:38 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp: In member function 'virtual void {anonymous}::test_if_constexpr::if_constexpr_returnTypeCanDiffer_withoutIdentityArg_Test::TestBody()': 
Dec 16 10:01:38 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:113:71: error: expected ',' before ')' token 
Dec 16 10:01:38      static_assert(std::is_same<std::string, decltype(a_string)>::value); 
Dec 16 10:01:38                                                                        ^ 
Dec 16 10:01:38 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:113:71: error: expected string-literal before ')' token 
Dec 16 10:01:38 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected ',' before ')' token 
Dec 16 10:01:38      static_assert(std::is_same<int64_t, decltype(an_int)>::value); 
Dec 16 10:01:38                                                                  ^ 
Dec 16 10:01:38 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected string-literal before ')' token 
Dec 16 10:01:38 c10/test/CMakeFiles/c10_C++17_test.dir/build.make:62: recipe for target 'c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o' failed 
Dec 16 10:01:38 make[2]: *** [c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o] Error 1 
Dec 16 10:01:38 CMakeFiles/Makefile2:2595: recipe for target 'c10/test/CMakeFiles/c10_C++17_test.dir/all' failed 
Dec 16 10:01:38 make[1]: *** [c10/test/CMakeFiles/c10_C++17_test.dir/all] Error 2 
Dec 16 10:01:38 make[1]: *** Waiting for unfinished jobs.... 
Dec 16 10:01:39 [ 41%] Generating src/x86_64-fma/blas/s4c6gemm.py.o 
Dec 16 10:01:39 [ 41%] Generating src/x86_64-fma/blas/conv1x1.py.o 
Dec 16 10:01:40 [ 41%] Generating src/x86_64-fma/blas/sgemm.py.o 
Dec 16 10:01:40 [ 41%] Generating src/x86_64-fma/max-pooling.py.o 
Dec 16 10:01:40 [ 41%] Generating src/x86_64-fma/relu.py.o 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3962130) pytorch_linux_xenial_py3_6_gcc5_4_build (2/5)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74952439))
<details>
<summary>
<code>Dec 16 10:03:20 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected string-literal before ')' token</code>
</summary>

```
Dec 16 10:03:20                                                                  ^ 
Dec 16 10:03:20 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:105:65: error: expected string-literal before ')' token 
Dec 16 10:03:20 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp: In member function 'virtual void {anonymous}::test_if_constexpr::if_constexpr_returnTypeCanDiffer_withoutIdentityArg_Test::TestBody()': 
Dec 16 10:03:20 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:113:71: error: expected ',' before ')' token 
Dec 16 10:03:20      static_assert(std::is_same<std::string, decltype(a_string)>::value); 
Dec 16 10:03:20                                                                        ^ 
Dec 16 10:03:20 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:113:71: error: expected string-literal before ')' token 
Dec 16 10:03:20 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected ',' before ')' token 
Dec 16 10:03:20      static_assert(std::is_same<int64_t, decltype(an_int)>::value); 
Dec 16 10:03:20                                                                  ^ 
Dec 16 10:03:20 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected string-literal before ')' token 
Dec 16 10:03:20 c10/test/CMakeFiles/c10_C++17_test.dir/build.make:62: recipe for target 'c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o' failed 
Dec 16 10:03:20 make[2]: *** [c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o] Error 1 
Dec 16 10:03:20 make[1]: *** [c10/test/CMakeFiles/c10_C++17_test.dir/all] Error 2 
Dec 16 10:03:20 make[1]: *** Waiting for unfinished jobs.... 
Dec 16 10:03:20 CMakeFiles/Makefile2:2396: recipe for target 'c10/test/CMakeFiles/c10_C++17_test.dir/all' failed 
Dec 16 10:03:21 [ 40%] Linking CXX executable ../../bin/c10_typeid_test 
Dec 16 10:03:21 [ 40%] Built target c10_typeid_test 
Dec 16 10:03:21 [ 40%] Linking CXX executable ../../bin/c10_DeviceGuard_test 
Dec 16 10:03:21 [ 40%] Built target c10_DeviceGuard_test 
Dec 16 10:03:21 make: *** [all] Error 2 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3962139) binary_linux_libtorch_2_7m_cpu_gcc5_4_cxx11-abi_shared-with-deps_build (3/5)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74954615))
<details>
<summary>
<code>Dec 16 10:02:46 FAILED: c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o</code>
</summary>

```
Dec 16 10:02:45 [750/2999] Building CXX object c10/test/CMakeFiles/c10_typeid_test.dir/util/typeid_test.cpp.o 
Dec 16 10:02:45 [751/2999] Building CXX object c10/test/CMakeFiles/c10_tempfile_test.dir/util/tempfile_test.cpp.o 
Dec 16 10:02:45 [752/2999] Building CXX object c10/test/CMakeFiles/c10_LeftRight_test.dir/util/LeftRight_test.cpp.o 
Dec 16 10:02:45 [753/2999] Linking CXX executable bin/c10_tempfile_test 
Dec 16 10:02:45 [754/2999] Linking CXX executable bin/c10_typeid_test 
Dec 16 10:02:45 [755/2999] Linking CXX executable bin/c10_LeftRight_test 
Dec 16 10:02:45 [756/2999] Building CXX object c10/test/CMakeFiles/c10_string_view_test.dir/util/string_view_test.cpp.o 
Dec 16 10:02:45 [757/2999] Linking CXX executable bin/c10_string_view_test 
Dec 16 10:02:46 [758/2999] Building CXX object c10/test/CMakeFiles/c10_TensorTypeSet_test.dir/core/TensorTypeSet_test.cpp.o 
Dec 16 10:02:46 [759/2999] Building CXX object c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o 
Dec 16 10:02:46 FAILED: c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o  
ror=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -O3  -fPIE   -DCAFFE2_USE_GLOO -DHAVE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD -std=gnu++14 -MD -MT c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o -MF c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o.d -o c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o -c ../c10/test/util/C++17_test.cpp 
Dec 16 10:02:46 ../c10/test/util/C++17_test.cpp: In member function ‘virtual void {anonymous}::test_if_constexpr::if_constexpr_returnTypeCanDiffer_withIdentityArg_Test::TestBody()’: 
Dec 16 10:02:46 ../c10/test/util/C++17_test.cpp:99:71: error: expected ‘,’ before ‘)’ token 
Dec 16 10:02:46      static_assert(std::is_same<std::string, decltype(a_string)>::value); 
Dec 16 10:02:46                                                                        ^ 
Dec 16 10:02:46 ../c10/test/util/C++17_test.cpp:99:71: error: expected string-literal before ‘)’ token 
Dec 16 10:02:46 ../c10/test/util/C++17_test.cpp:105:65: error: expected ‘,’ before ‘)’ token 
Dec 16 10:02:46      static_assert(std::is_same<int64_t, decltype(an_int)>::value); 
Dec 16 10:02:46                                                                  ^ 
Dec 16 10:02:46 ../c10/test/util/C++17_test.cpp:105:65: error: expected string-literal before ‘)’ token 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3962122) pytorch_windows_build (4/5)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74959535))
<details>
<summary>
<code>FAILED: c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.obj</code>
</summary>

```

m\hip\include -I\include -I..\cmake\..\third_party\cub -Iwin_tmp\magma\include -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\include" /DWIN32 /D_WINDOWS /GR  /w /EHa /MP /bigobj -openmp:experimental /MD /O2 /Ob2  /w /EHa /MP /bigobj   -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD -DC10_BUILD_MAIN_LIB -std:c++14 /showIncludes /Foc10\CMakeFiles\c10.dir\core\UndefinedTensorImpl.cpp.obj /Fdc10\CMakeFiles\c10.dir\ /FS -c ..\c10\core\UndefinedTensorImpl.cpp 
Microsoft (R) C/C++ Optimizing Compiler Version 19.23.28107 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

 -I..\third_party\googletest\googlemock -I..\third_party\googletest\googletest\include -I..\third_party\googletest\googletest /DWIN32 /D_WINDOWS /GR  /w /EHa /MP /bigobj -openmp:experimental /MD /O2 /Ob2  /w /EHa /MP /bigobj   -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD -std:c++14 /showIncludes /Foc10\test\CMakeFiles\c10_StreamGuard_test.dir\core\StreamGuard_test.cpp.obj /Fdc10\test\CMakeFiles\c10_StreamGuard_test.dir\ /FS -c ..\c10\test\core\StreamGuard_test.cpp 
Microsoft (R) C/C++ Optimizing Compiler Version 19.23.28107 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

etest\googlemock\include -I..\third_party\googletest\googlemock -I..\third_party\googletest\googletest\include -I..\third_party\googletest\googletest /DWIN32 /D_WINDOWS /GR  /w /EHa /MP /bigobj -openmp:experimental /MD /O2 /Ob2  /w /EHa /MP /bigobj   -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD -std:c++14 /showIncludes /Foc10\test\CMakeFiles\c10_C++17_test.dir\util\C++17_test.cpp.obj /Fdc10\test\CMakeFiles\c10_C++17_test.dir\ /FS -c ..\c10\test\util\C++17_test.cpp 
FAILED: c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.obj  
etest\googlemock\include -I..\third_party\googletest\googlemock -I..\third_party\googletest\googletest\include -I..\third_party\googletest\googletest /DWIN32 /D_WINDOWS /GR  /w /EHa /MP /bigobj -openmp:experimental /MD /O2 /Ob2  /w /EHa /MP /bigobj   -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD -std:c++14 /showIncludes /Foc10\test\CMakeFiles\c10_C++17_test.dir\util\C++17_test.cpp.obj /Fdc10\test\CMakeFiles\c10_C++17_test.dir\ /FS -c ..\c10\test\util\C++17_test.cpp 
..\c10\test\util\C++17_test.cpp(99): error C2429: language feature 'terse static assert' requires compiler flag '/std:c++17'
..\c10\test\util\C++17_test.cpp(105): error C2429: language feature 'terse static assert' requires compiler flag '/std:c++17'
..\c10\test\util\C++17_test.cpp(113): error C2429: language feature 'terse static assert' requires compiler flag '/std:c++17'
..\c10\test\util\C++17_test.cpp(119): error C2429: language feature 'terse static assert' requires compiler flag '/std:c++17'
Microsoft (R) C/C++ Optimizing Compiler Version 19.23.28107 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

include -I\opt\rocm\hip\include -I\include -I..\cmake\..\third_party\cub -Iwin_tmp\magma\include -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\include" /DWIN32 /D_WINDOWS /GR  /w /EHa /MP /bigobj -openmp:experimental /MD /O2 /Ob2  /w /EHa /MP /bigobj   -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD -DC10_BUILD_MAIN_LIB -std:c++14 /showIncludes /Foc10\CMakeFiles\c10.dir\core\TensorImpl.cpp.obj /Fdc10\CMakeFiles\c10.dir\ /FS -c ..\c10\core\TensorImpl.cpp 
Microsoft (R) C/C++ Optimizing Compiler Version 19.23.28107 for x64
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3962123) pytorch_linux_xenial_cuda9_cudnn7_py3_build (5/5)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74961823))
<details>
<summary>
<code>Dec 16 10:06:37 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected string-literal before ')' token</code>
</summary>

```
Dec 16 10:06:37                                                                  ^ 
Dec 16 10:06:37 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:105:65: error: expected string-literal before ')' token 
Dec 16 10:06:37 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp: In member function 'virtual void {anonymous}::test_if_constexpr::if_constexpr_returnTypeCanDiffer_withoutIdentityArg_Test::TestBody()': 
Dec 16 10:06:37 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:113:71: error: expected ',' before ')' token 
Dec 16 10:06:37      static_assert(std::is_same<std::string, decltype(a_string)>::value); 
Dec 16 10:06:37                                                                        ^ 
Dec 16 10:06:37 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:113:71: error: expected string-literal before ')' token 
Dec 16 10:06:37 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected ',' before ')' token 
Dec 16 10:06:37      static_assert(std::is_same<int64_t, decltype(an_int)>::value); 
Dec 16 10:06:37                                                                  ^ 
Dec 16 10:06:37 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected string-literal before ')' token 
Dec 16 10:06:37 c10/test/CMakeFiles/c10_C++17_test.dir/build.make:62: recipe for target 'c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o' failed 
Dec 16 10:06:37 make[2]: *** [c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o] Error 1 
Dec 16 10:06:37 CMakeFiles/Makefile2:2736: recipe for target 'c10/test/CMakeFiles/c10_C++17_test.dir/all' failed 
Dec 16 10:06:37 make[1]: *** [c10/test/CMakeFiles/c10_C++17_test.dir/all] Error 2 
Dec 16 10:06:37 make[1]: *** Waiting for unfinished jobs.... 
Dec 16 10:06:37 Compiling  all_reduce.cu                       > /var/lib/jenkins/workspace/build/nccl/obj/collectives/device/all_reduce_min_u32.o 
Dec 16 10:06:42 Compiling  all_reduce.cu                       > /var/lib/jenkins/workspace/build/nccl/obj/collectives/device/all_reduce_min_i64.o 
Dec 16 10:06:46 Compiling  all_reduce.cu                       > /var/lib/jenkins/workspace/build/nccl/obj/collectives/device/all_reduce_min_u64.o 
Dec 16 10:06:50 Compiling  all_reduce.cu                       > /var/lib/jenkins/workspace/build/nccl/obj/collectives/device/all_reduce_min_f16.o 
Dec 16 10:06:50 [ 35%] Linking CXX executable ../../bin/c10_intrusive_ptr_test 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

smessmer committed:
Update on "Reimplement c10::apply / c10::invoke with c++14"


These had an incomplete implementation before because the one from cppreference.com didn't work in C++11.
Now with C++14, we can make the full implementation from cppreference.com work and have c++17-compatible verisons of std::apply/std::invoke.

Differential Revision: [D18869642](https://our.internmc.facebook.com/intern/diff/D18869642/)

[ghstack-poisoned]

smessmer committed:
Update on "Reimplement c10::apply / c10::invoke with c++14"


These had an incomplete implementation before because the one from cppreference.com didn't work in C++11.
Now with C++14, we can make the full implementation from cppreference.com work and have c++17-compatible verisons of std::apply/std::invoke.

Differential Revision: [D18869642](https://our.internmc.facebook.com/intern/diff/D18869642/)

[ghstack-poisoned]

smessmer committed:
Update on "Reimplement c10::apply / c10::invoke with c++14"


These had an incomplete implementation before because the one from cppreference.com didn't work in C++11.
Now with C++14, we can make the full implementation from cppreference.com work and have c++17-compatible verisons of std::apply/std::invoke.

Differential Revision: [D18869642](https://our.internmc.facebook.com/intern/diff/D18869642/)

[ghstack-poisoned]

smessmer committed:
Update on "Reimplement c10::apply / c10::invoke with c++14"


These had an incomplete implementation before because the one from cppreference.com didn't work in C++11.
Now with C++14, we can make the full implementation from cppreference.com work and have c++17-compatible verisons of std::apply/std::invoke.

Differential Revision: [D18869642](https://our.internmc.facebook.com/intern/diff/D18869642/)

[ghstack-poisoned]

smessmer committed:
Update on "Reimplement c10::apply / c10::invoke with c++14"


These had an incomplete implementation before because the one from cppreference.com didn't work in C++11.
Now with C++14, we can make the full implementation from cppreference.com work and have c++17-compatible verisons of std::apply/std::invoke.

Differential Revision: [D18869642](https://our.internmc.facebook.com/intern/diff/D18869642/)

[ghstack-poisoned]

smessmer committed:
Update on "Reimplement c10::apply / c10::invoke with c++14"


These had an incomplete implementation before because the one from cppreference.com didn't work in C++11.
Now with C++14, we can make the full implementation from cppreference.com work and have c++17-compatible verisons of std::apply/std::invoke.

Differential Revision: [D18869642](https://our.internmc.facebook.com/intern/diff/D18869642/)

[ghstack-poisoned]

smessmer committed:
Update on "Reimplement c10::apply / c10::invoke with c++14"


These had an incomplete implementation before because the one from cppreference.com didn't work in C++11.
Now with C++14, we can make the full implementation from cppreference.com work and have c++17-compatible verisons of std::apply/std::invoke.

Differential Revision: [D18869642](https://our.internmc.facebook.com/intern/diff/D18869642/)

[ghstack-poisoned]

smessmer committed:
Update on "Reimplement c10::apply / c10::invoke with c++14"


These had an incomplete implementation before because the one from cppreference.com didn't work in C++11.
Now with C++14, we can make the full implementation from cppreference.com work and have c++17-compatible verisons of std::apply/std::invoke.

Differential Revision: [D18869642](https://our.internmc.facebook.com/intern/diff/D18869642/)

[ghstack-poisoned]

smessmer committed:
Update on "Reimplement c10::apply / c10::invoke with c++14"


These had an incomplete implementation before because the one from cppreference.com didn't work in C++11.
Now with C++14, we can make the full implementation from cppreference.com work and have c++17-compatible verisons of std::apply/std::invoke.

Differential Revision: [D18869642](https://our.internmc.facebook.com/intern/diff/D18869642/)

[ghstack-poisoned]

smessmer committed:
Update on "Reimplement c10::apply / c10::invoke with c++14"


These had an incomplete implementation before because the one from cppreference.com didn't work in C++11.
Now with C++14, we can make the full implementation from cppreference.com work and have c++17-compatible verisons of std::apply/std::invoke.

Differential Revision: [D18869642](https://our.internmc.facebook.com/intern/diff/D18869642/)

[ghstack-poisoned]

smessmer committed:
Update on "Reimplement c10::apply / c10::invoke with c++14"


These had an incomplete implementation before because the one from cppreference.com didn't work in C++11.
Now with C++14, we can make the full implementation from cppreference.com work and have c++17-compatible verisons of std::apply/std::invoke.

Differential Revision: [D18869642](https://our.internmc.facebook.com/intern/diff/D18869642/)

[ghstack-poisoned]

smessmer committed:
Update on "Reimplement c10::apply / c10::invoke with c++14"


These had an incomplete implementation before because the one from cppreference.com didn't work in C++11.
Now with C++14, we can make the full implementation from cppreference.com work and have c++17-compatible verisons of std::apply/std::invoke.

Differential Revision: [D18869642](https://our.internmc.facebook.com/intern/diff/D18869642/)

[ghstack-poisoned]

smessmer committed:
Update on "Reimplement c10::apply / c10::invoke with c++14"


These had an incomplete implementation before because the one from cppreference.com didn't work in C++11.
Now with C++14, we can make the full implementation from cppreference.com work and have c++17-compatible verisons of std::apply/std::invoke.

Differential Revision: [D18869642](https://our.internmc.facebook.com/intern/diff/D18869642/)

[ghstack-poisoned]

smessmer committed:
Update on "Reimplement c10::apply / c10::invoke with c++14"


These had an incomplete implementation before because the one from cppreference.com didn't work in C++11.
Now with C++14, we can make the full implementation from cppreference.com work and have c++17-compatible verisons of std::apply/std::invoke.

Differential Revision: [D18869642](https://our.internmc.facebook.com/intern/diff/D18869642/)

[ghstack-poisoned]
